utils/bash-completion/bluealsa

```diff
@@ -0,0 +1,375 @@
+			return
+			;;
+		--sbc-quality)
+			COMPREPLY=( 0 1 2 3 )
+			return
+			;;
+		--mp3-quality|--mp3-vbr-quality)
+			COMPREPLY=( 0 1 2 3 4 5 6 7 8 9 )
+			return
+			;;
+		--aac-latm-version)
+			COMPREPLY=( 0 1 )
+			return
+			;;
+		--aac-vbr-mode)
+			COMPREPLY=( 0 1 2 3 4 5 )
+			return
+			;;
+		--ldac-eqmid)
+			COMPREPLY=( 0 1 2 )
+			return
+			;;
+		--xapl-resp-name)
+			COMPREPLY=( $(compgen -W "BlueALSA iPhone" -- $cur) )
+			return
+			;;
+		--*)
+			[[ ${COMP_WORDS[COMP_CWORD]} = = ]] && return
+			;;
+	esac
+
+	_bluealsa_complete_options $1
+}
+
+# completion function for bluealsa-aplay
+# completes available dbus suffices and ALSA pcms in addition to options
+# - does not complete MAC addresses
+# requires aplay to list ALSA pcms
+_bluealsa_aplay() {
+	local cur prev words cword split list
+
+	_init_completion -s -n : || return
+
+	case "$prev" in
+		--dbus|-B)
+			_have dbus-send || return
+			list=$(_bluealsa_list_dbus_suffices)
+			COMPREPLY=( $(compgen -W "$list" -- $cur) )
+			return
+			;;
+		--dbus=)
+			return
+			;;
+		--pcm|-D)
+			_have aplay || return
+			COMPREPLY=( $(compgen -W "$(_bluealsa_alsa_pcms)" -- $cur) )
+			# ALSA pcm names can contain '=' and ':', both of which cause
+			# problems for bash completion if it considers them to be word
+			# terminators. So we adjust the list of candidate matches to allow
+			# for this.
+		    if [[ "$cur" == *=* && "$COMP_WORDBREAKS" == *=* ]]; then
+				# Remove equal-word prefix from COMPREPLY items
+				local equal_prefix=${cur%"${cur##*=}"}
+				local i=${#COMPREPLY[*]}
+				while [[ $((--i)) -ge 0 ]]; do
+				    COMPREPLY[$i]=${COMPREPLY[$i]#"$equal_prefix"}
+				done
+			fi
+			__ltrim_colon_completions "$cur"
+			return
+			;;
+		--*)
+			[[ ${COMP_WORDS[COMP_CWORD]} = = ]] && return
+			;;
+	esac
+
+	_bluealsa_complete_options $1
+}
+
+# completion function for bluealsa-cli
+# complete available dbus suffices, command names and pcm paths in addition to options
+_bluealsa_cli() {
+	local cur prev words cword split
+	local dbus_opt nonopt_offset service
+
+	_init_completion -s || return
+
+	_bluealsa_util_init $1 || return
+
+	# get the command names supported by this version of bluealsa-cli
+	local simple_commands="$(_bluealsa_cli_simple_commands $1)"
+	local path_commands="$(_bluealsa_cli_path_commands $1)"
+
+	case "$prev" in
+		--dbus|-B)
+			_have dbus-send || return
+			list=$(_bluealsa_list_dbus_suffices)
+			COMPREPLY=( $(compgen -W "$list" -- $cur) )
+			return
+			;;
+		--dbus=)
+			return
+			;;
+		--*)
+			[[ ${COMP_WORDS[COMP_CWORD]} = = ]] && return
+			;;
+	esac
+
+	case "$nonopt_offset" in
+		"$COMP_CWORD")
+			# list available commands
+			COMPREPLY=( $(compgen -W "$simple_commands $path_commands" -- "$cur") )
+			return
+			;;
+		$((COMP_CWORD - 1)))
+			# if previous word was command then list available paths
+			if [[ "$path_commands" =~ "$prev" ]]; then
+				COMPREPLY=( $(compgen -W "$("$1" $dbus_opt list-pcms 2>/dev/null)" -- $cur) )
```

Comment:
In line 297 the $dbus_opt is in quotations, but here it's not.

Reply:
Yes line 288 is invoking `bluealsa-cli`, so putting an empty string as first arg would cause "invalid command" error. Line 297 is invoking function _bluealsa_codecs which requires the second argument, even if is just the empty string.